File: frontend/src/schemas/insight.ts
import { z } from "zod";

// Insight schema
export const InsightSchema = z.object({
	id: z
		.string()
		.uuid()
		.optional()
		.default(() => crypto.randomUUID()),
	observation: z.string(),
	severity: z.enum(["low", "medium", "high", "critical"]).optional().nullable(),
	category: z
		.enum(["performance", "security", "availability", "anomaly"])
		.optional()
		.nullable(),
	related_logs: z.array(z.string().uuid()).optional().nullable(),
	timestamp: z
		.string()
		.optional()
		.default(() => new Date().toISOString()),
	recommendation: z.string().optional().nullable(),
	confidence: z.number().min(0).max(1).optional().nullable(),
	tags: z.array(z.string()).optional().nullable(),
});

export type Insight = z.infer<typeof InsightSchema>;

// InsightList schema
export const InsightListSchema = z.object({
	insights: z.array(InsightSchema),
});

export type InsightList = z.infer<typeof InsightListSchema>;

File: backend/models/insight.py
import uuid

from datetime import datetime

from pydantic import BaseModel, Field


class Insight(BaseModel):
    """Insight model."""
    id: str = Field(
        default_factory=lambda: str(uuid.uuid4()),
        description="Unique identifier for the insight"
    )
    observation: str = Field(
        ...,
        description="Human-readable AI observation"
    )
    severity: str | None = Field(
        None,
        description="Severity: low, medium, high, critical"
    )
    category: str | None = Field(
        None,
        description="Category: performance, security, availability, anomaly"
    )
    related_logs: list[str] | None = Field(
        None,
        description="List of log IDs that led to this insight"
    )
    timestamp: datetime = Field(
        default_factory=datetime.utcnow,
        description="Insight creation time"
    )
    recommendation: str | None = Field(
        None,
        description="Suggested action"
    )
    confidence: float | None = Field(
        None,
        description="AI confidence score (0–1)"  # noqa: RUF001
    )
    tags: list[str] | None = Field(
        None,
        description="List of tags for categorization"
    )


class InsightList(BaseModel):
    """Insight List Model."""
    insights: list[Insight]
